src/ray/core_worker/pubsub/pubsub_coordinator.cc

```diff
@@ -0,0 +1,206 @@
+}
+
+absl::optional<std::reference_wrapper<const absl::flat_hash_set<NodeID>>>
+SubscriptionIndex::GetSubscriberIdsByObjectId(const ObjectID &object_id) const {
+  auto it = objects_to_subscribers_.find(object_id);
+  if (it == objects_to_subscribers_.end()) {
+    return absl::nullopt;
+  }
+  return absl::optional<std::reference_wrapper<const absl::flat_hash_set<NodeID>>>{
+      std::ref(it->second)};
+}
+
+bool SubscriptionIndex::HasObjectId(const ObjectID &object_id) const {
+  return objects_to_subscribers_.count(object_id);
+}
+
+bool SubscriptionIndex::HasSubscriber(const NodeID &subscriber_id) const {
+  return subscribers_to_objects_.count(subscriber_id);
+}
+
+bool SubscriptionIndex::EraseSubscriber(const NodeID &subscriber_id) {
+  auto subscribing_objects_it = subscribers_to_objects_.find(subscriber_id);
+  if (subscribing_objects_it == subscribers_to_objects_.end()) {
+    return false;
+  }
+
+  auto &subscribing_objects = subscribing_objects_it->second;
+  for (const auto &object_id : subscribing_objects) {
+    // Erase the subscriber from the object map.
+    auto subscribers_it = objects_to_subscribers_.find(object_id);
+    if (subscribers_it == objects_to_subscribers_.end()) {
+      continue;
+    }
+    auto &subscribers = subscribers_it->second;
+    subscribers.erase(subscriber_id);
+    if (subscribers.size() == 0) {
+      objects_to_subscribers_.erase(subscribers_it);
+    }
+  }
+  subscribers_to_objects_.erase(subscribing_objects_it);
+  return true;
+}
+
+bool SubscriptionIndex::EraseEntry(const ObjectID &object_id,
+                                   const NodeID &subscriber_id) {
+  // Erase from subscribers_to_objects_;
+  auto subscribers_to_objects_it = subscribers_to_objects_.find(subscriber_id);
+  if (subscribers_to_objects_it == subscribers_to_objects_.end()) {
+    return false;
+  }
+  auto &objects = subscribers_to_objects_it->second;
+  auto object_it = objects.find(object_id);
+  if (object_it == objects.end()) {
+    RAY_CHECK(objects_to_subscribers_.count(object_id) == 0);
+    return false;
+  }
+  objects.erase(object_it);
+  if (objects.size() == 0) {
+    subscribers_to_objects_.erase(subscribers_to_objects_it);
+  }
+
+  // Erase from objects_to_subscribers_.
+  auto objects_to_subscribers_it = objects_to_subscribers_.find(object_id);
+  // If code reaches this line, that means the object id was in the index.
+  RAY_CHECK(objects_to_subscribers_it != objects_to_subscribers_.end());
+  auto &subscribers = objects_to_subscribers_it->second;
+  auto subscriber_it = subscribers.find(subscriber_id);
+  // If code reaches this line, that means the subscriber id was in the index.
+  RAY_CHECK(subscriber_it != subscribers.end());
+  subscribers.erase(subscriber_it);
+  if (subscribers.size() == 0) {
+    objects_to_subscribers_.erase(objects_to_subscribers_it);
+  }
+  return true;
+}
+
+bool Subscriber::Connect(LongPollConnectCallback long_polling_reply_callback) {
```

Comment:
```suggestion
bool Subscriber::RegisterSubscriber(LongPollConnectCallback long_polling_reply_callback) {
```

Reply:
What about ConnectToSubscriber? Register sounds a little misleading imo because this can happen for every long polling request from the same subscriber.